File: sdks/python/src/ankka/testkit/unit.py
"""Unit testkits: one component, no sidecar, no network.

Effects are values, so a handler can be run against an in-memory state and its effect inspected.
Inputs, events, state and replies still round-trip through the component's own codecs, so a type
the default codec cannot encode fails here rather than on first deployment (the Scala testkit's
rule).
"""

from __future__ import annotations

import asyncio
import re
import typing
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

from ankka.client import ComponentClient
from ankka.context import CommandContext, Metadata, Principal, RequestContext
from ankka.effects.common import Error, Fail, NoReply, Reply, Retention
from ankka.endpoint import Endpoint, HttpProblem, RouteSpec
from ankka.event_sourced_entity import EventSourcedEntity, HandlerSpec

S = TypeVar("S")
E = TypeVar("E")


@dataclass(frozen=True)
class Materialised(Generic[S, E]):
    """What running an effect against a state produced: the shape the sidecar reduces to."""

    events: tuple[E, ...]
    new_state: S
    retention: Retention | None
    reply: Any
    error: Error | None

    @property
    def persisted(self) -> bool:
        return len(self.events) > 0

    def reply_or_raise(self) -> Any:
        if self.error is not None:
            raise AssertionError(f"the command was refused: {self.error.message} ({self.error.code.name})")
        return self.reply


def _run(coro: Any) -> Any:
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(coro)
    raise RuntimeError("the unit testkit is synchronous; call it outside a running event loop")


class _NoClient(ComponentClient):
    """A client that refuses: a unit test must not reach the platform."""

    def __init__(self) -> None:  # noqa: D401 - deliberately not calling super
        self.address = "unit-test"
        self._metadata = Metadata()

    def with_metadata(self, metadata: Metadata) -> ComponentClient:
        return self

    def _refuse(self) -> Any:
        raise RuntimeError("a unit test cannot call other components; use the integration testkit")

    def for_event_sourced_entity(self, component_id: str, entity_id: str) -> Any:
        return self._refuse()

    def for_key_value_entity(self, component_id: str, entity_id: str) -> Any:
        return self._refuse()

    def for_workflow(self, component_id: str, workflow_id: str) -> Any:
        return self._refuse()

    def for_agent(self, component_id: str, session_id: str) -> Any:
        return self._refuse()

    async def close(self) -> None:
        return None


class EventSourcedTestKit(Generic[S, E]):
    """Drives one event sourced entity instance through its handlers, folding events as it goes."""

    def __init__(self, entity_cls: type[EventSourcedEntity[S, E]], entity_id: str) -> None:
        self.entity_cls = entity_cls
        self.entity_id = entity_id
        self.entity = entity_cls()
        self.entity._bind(entity_id)
        self.state: S = self.entity.empty_state()
        self.sequence = 0
        self.deleted = False
        self.all_events: list[E] = []

    @classmethod
    def of(cls, entity_cls: type[EventSourcedEntity[S, E]], entity_id: str = "test") -> EventSourcedTestKit[S, E]:
        return cls(entity_cls, entity_id)

    def call(self, name: str, input: Any = None, metadata: Metadata | None = None) -> Materialised[S, E]:
        spec: HandlerSpec | None = self.entity_cls.handlers().get(name)
        if spec is None:
            raise AssertionError(f"{self.entity_cls.__name__} has no handler {name!r}; declared: {sorted(self.entity_cls.handlers())}")
        # Round-trip the input through the handler's codec, as the wire would.
        input_bytes = spec.input_codec.encode(input) if spec.input_type is not None else b""
        ctx = CommandContext(self.entity_id, self.entity_cls.component_id, metadata or Metadata(), self.sequence, _NoClient())
        effect = _run(self.entity._run(spec, self.state, input_bytes, ctx))
        if isinstance(effect.outcome, Fail):
            return Materialised((), self.state, None, None, effect.outcome.error)
        codec = self.entity_cls.event_codec
        events = tuple(codec.decode(codec.encode(ev)) for ev in effect.events)  # the wire's round trip
        new_state = self.state
        for ev in events:
            new_state = self.entity.apply_event(new_state, ev)
        reply: Any = None
        if isinstance(effect.outcome, Reply):
            value = effect.outcome.compute(new_state)
            reply = spec.reply_codec.decode(spec.reply_codec.encode(value))
        elif isinstance(effect.outcome, NoReply):
            reply = None
        # The state codec must round-trip too: a snapshot would.
        sc = self.entity_cls.state_codec
        new_state = sc.decode(sc.encode(new_state))
        self.state = new_state
        self.sequence += len(events)
        self.all_events.extend(events)
        return Materialised(events, new_state, effect.retention, reply, None)


_PLACEHOLDER = re.compile(r"\{[A-Za-z_][A-Za-z0-9_]*\}")


@dataclass(frozen=True)
class Response:
    status: int
    content_type: str
    body: bytes

    def text(self) -> str:
        return self.body.decode("utf-8")

    def json(self) -> Any:
        import json

        return json.loads(self.body.decode("utf-8"))


class EndpointTestKit:
    """Calls an endpoint's routes by path, binding parameters as the sidecar's router would."""

    def __init__(self, endpoint: Endpoint) -> None:
        self.endpoint = endpoint

    @classmethod
    def of(cls, endpoint_cls: type[Endpoint], *args: Any, **kwargs: Any) -> EndpointTestKit:
        return cls(endpoint_cls(*args, **kwargs))

    def _match(self, method: str, path: str) -> tuple[RouteSpec, list[str]] | None:
        prefix = type(self.endpoint).prefix
        if not path.startswith(prefix):
            return None
        rest = path[len(prefix):] or "/"
        candidates: list[tuple[int, RouteSpec, list[str]]] = []
        for spec in type(self.endpoint).routes().values():
            if spec.method != method.upper():
                continue
            pattern = "^" + _PLACEHOLDER.sub("([^/]+)", re.escape(spec.template).replace("\\{", "{").replace("\\}", "}")) + "$"
            m = re.match(pattern, rest)
            if m:
                literal_segments = sum(1 for seg in spec.template.split("/") if seg and not seg.startswith("{"))
                candidates.append((literal_segments, spec, list(m.groups())))
        if not candidates:
            return None
        candidates.sort(key=lambda c: -c[0])  # literal segments outrank parameters
        _, spec, args = candidates[0]
        return spec, args

    def request(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: list[tuple[str, str]] | None = None,
        headers: list[tuple[str, str]] | None = None,
        principal: Principal | None = None,
    ) -> Response:
        matched = self._match(method, path)
        if matched is None:
            return Response(404, "text/plain", f"no route {method} {path}".encode())
        spec, args = matched
        body_bytes = b""
        if body is not None and spec.body_codec is not None:
            body_bytes = body if isinstance(body, bytes) else spec.body_codec.encode(body)
        ctx = RequestContext(tuple(query or ()), tuple(headers or ()), principal, Metadata())
        try:
            if spec.streaming:
                frames = _run(self._collect(spec, args, body_bytes, ctx))
                return Response(200, "text/event-stream", "\n".join(frames).encode("utf-8"))
            status, content_type, out = _run(self.endpoint._handle(spec, args, body_bytes, ctx))
            return Response(status, content_type, out)
        except HttpProblem as p:
            return Response(p.status, "text/plain", p.message.encode("utf-8"))

    async def _collect(self, spec: RouteSpec, args: list[str], body: bytes, ctx: RequestContext) -> list[str]:
        return [frame async for frame in self.endpoint._handle_stream(spec, args, body, ctx)]

    def get(self, path: str, **kwargs: Any) -> Response:
        return self.request("GET", path, **kwargs)

    def post(self, path: str, body: Any = None, **kwargs: Any) -> Response:
        return self.request("POST", path, body=body, **kwargs)

    def put(self, path: str, body: Any = None, **kwargs: Any) -> Response:
        return self.request("PUT", path, body=body, **kwargs)

    def delete(self, path: str, **kwargs: Any) -> Response:
        return self.request("DELETE", path, **kwargs)


# ── The other kinds ──────────────────────────────────────────────────────────


from ankka.consumer import Consumer  # noqa: E402
from ankka.effects.consumer import ConsumerEffect  # noqa: E402
from ankka.effects.key_value import KeyValueEffect  # noqa: E402
from ankka.effects.timed_action import TimedActionEffect  # noqa: E402
from ankka.effects.view import ViewEffect  # noqa: E402
from ankka.effects.workflow import StepOutcome, StepRef, TransitionTo, WorkflowStepEffect  # noqa: E402
from ankka.key_value_entity import KeyValueEntity  # noqa: E402
from ankka.timed_action import TimedAction  # noqa: E402
from ankka.view import View  # noqa: E402
from ankka.workflow import Workflow  # noqa: E402

Row = TypeVar("Row")


@dataclass(frozen=True)
class KeyValueMaterialised(Generic[S]):
    new_state: S
    retention: Retention | None
    reply: Any
    error: Error | None
    changed: bool


class KeyValueTestKit(Generic[S]):
    def __init__(self, entity_cls: type[KeyValueEntity[S]], entity_id: str) -> None:
        self.entity_cls = entity_cls
        self.entity = entity_cls()
        self.entity.__class__  # noqa: B018 - keeps mypy honest about the generic
        self.entity._bind(entity_id)
        self.entity_id = entity_id
        self.state: S = self.entity.empty_state()

    @classmethod
    def of(cls, entity_cls: type[KeyValueEntity[S]], entity_id: str = "test") -> KeyValueTestKit[S]:
        return cls(entity_cls, entity_id)

    def call(self, name: str, input: Any = None) -> KeyValueMaterialised[S]:
        spec = self.entity_cls.handlers().get(name)
        if spec is None:
            raise AssertionError(f"{self.entity_cls.__name__} has no handler {name!r}")
        input_bytes = spec.input_codec.encode(input) if spec.input_type is not None else b""
        ctx = CommandContext(self.entity_id, self.entity_cls.component_id, Metadata(), 0, _NoClient())
        effect: KeyValueEffect[S, Any] = _run(self.entity._run(spec, self.state, input_bytes, ctx))
        if isinstance(effect.outcome, Fail):
            return KeyValueMaterialised(self.state, None, None, effect.outcome.error, False)
        sc = self.entity_cls.state_codec
        new_state = self.state if effect.new_state is None else sc.decode(sc.encode(effect.new_state))
        reply: Any = None
        if isinstance(effect.outcome, Reply):
            reply = spec.reply_codec.decode(spec.reply_codec.encode(effect.outcome.compute(new_state)))
        changed = effect.new_state is not None
        self.state = new_state
        return KeyValueMaterialised(new_state, effect.retention, reply, None, changed)


@dataclass(frozen=True)
class WorkflowMaterialised(Generic[S]):
    new_state: S
    transition: StepRef | None
    reply: Any
    error: Error | None


@dataclass(frozen=True)
class StepMaterialised(Generic[S]):
    new_state: S
    next: StepOutcome


class WorkflowTestKit(Generic[S]):
    """Drives commands and steps by hand: ``call`` runs a command, ``run_step`` a step, and
    ``run_until_end`` follows transitions (not pauses) until the workflow ends."""

    def __init__(self, workflow_cls: type[Workflow[S]], workflow_id: str) -> None:
        self.workflow_cls = workflow_cls
        self.workflow = workflow_cls()
        self.workflow._bind(workflow_id)
        self.workflow_id = workflow_id
        self.state: S = self.workflow.empty_state()
        self.pending: StepRef | None = None
        self.transitions: list[str] = []

    @classmethod
    def of(cls, workflow_cls: type[Workflow[S]], workflow_id: str = "test") -> WorkflowTestKit[S]:
        return cls(workflow_cls, workflow_id)

    def call(self, name: str, input: Any = None) -> WorkflowMaterialised[S]:
        spec = self.workflow_cls.handlers().get(name)
        if spec is None:
            raise AssertionError(f"{self.workflow_cls.__name__} has no handler {name!r}")
        input_bytes = spec.input_codec.encode(input) if spec.input_type is not None else b""
        ctx = CommandContext(self.workflow_id, self.workflow_cls.component_id, Metadata(), 0, _NoClient())
        effect = _run(self.workflow._run(spec, self.state, input_bytes, ctx))
        if isinstance(effect.outcome, Fail):
            return WorkflowMaterialised(self.state, None, None, effect.outcome.error)
        new_state = self.state if effect.new_state is None else effect.new_state
        reply: Any = None
        if isinstance(effect.outcome, Reply):
            reply = spec.reply_codec.decode(spec.reply_codec.encode(effect.outcome.compute(new_state)))
        self.state = new_state
        if effect.transition is not None:
            self.pending = effect.transition
            self.transitions.append(effect.transition.step)
        return WorkflowMaterialised(new_state, effect.transition, reply, None)

    def run_step(self, step: str | None = None, input: Any = None) -> StepMaterialised[S]:
        ref = StepRef(step, input) if step is not None else self.pending
        if ref is None:
            raise AssertionError("no step is pending; name one")
        spec = self.workflow_cls.steps().get(ref.step)
        if spec is None:
            raise AssertionError(f"{self.workflow_cls.__name__} has no step {ref.step!r}")
        input_bytes = spec.input_codec.encode(ref.input) if spec.input_type is not None else None
        ctx = CommandContext(self.workflow_id, self.workflow_cls.component_id, Metadata(), 0, _NoClient())
        effect: WorkflowStepEffect[S] = _run(self.workflow._run_step(spec, self.state, input_bytes, ctx))
        if effect.new_state is not None:
            self.state = effect.new_state
        self.pending = effect.next.ref if isinstance(effect.next, TransitionTo) else None
        if self.pending is not None:
            self.transitions.append(self.pending.step)
        return StepMaterialised(self.state, effect.next)

    def run_until_end(self, limit: int = 100) -> StepOutcome:
        last: StepOutcome | None = None
        for _ in range(limit):
            if self.pending is None:
                break
            last = self.run_step().next
        if last is None:
            raise AssertionError("nothing to run")
        return last


class ViewTestKit(Generic[Row]):
    """Feeds events to a view, keeping one row per key as the sidecar's projection would."""

    def __init__(self, view_cls: type[View[Any, Row]]) -> None:
        self.view_cls = view_cls
        self.rows: dict[str, Row] = {}

    @classmethod
    def of(cls, view_cls: type[View[Any, Row]]) -> ViewTestKit[Row]:
        return cls(view_cls)

    def on_change(self, key: str, event: Any) -> ViewEffect:
        view = self.view_cls()
        ec, rc = self.view_cls.event_codec, self.view_cls.row_codec
        current = self.rows.get(key)
        effect = _run(view._handle(ec.encode(event), rc.encode(current) if current is not None else None, Metadata()))
        from ankka.effects.view import DeleteRow, UpdateRow

        if isinstance(effect, UpdateRow):
            self.rows[key] = rc.decode(rc.encode(effect.row))
        elif isinstance(effect, DeleteRow):
            self.rows.pop(key, None)
        return typing.cast(ViewEffect, effect)

    def get(self, key: str) -> Row | None:
        return self.rows.get(key)


class ConsumerTestKit:
    def __init__(self, consumer_cls: type[Consumer[Any, Any]]) -> None:
        self.consumer_cls = consumer_cls
        self.produced: list[Any] = []

    @classmethod
    def of(cls, consumer_cls: type[Consumer[Any, Any]]) -> ConsumerTestKit:
        return cls(consumer_cls)

    def on_message(self, message: Any) -> ConsumerEffect:
        consumer = self.consumer_cls(_NoClient())
        mc = self.consumer_cls.message_codec
        effect = _run(consumer._handle(mc.encode(message), Metadata()))
        from ankka.effects.consumer import Produce

        if isinstance(effect, Produce):
            oc = self.consumer_cls.out_codec
            assert oc is not None
            self.produced.append(oc.decode(oc.encode(effect.payload)))
        return typing.cast(ConsumerEffect, effect)


class TimedActionTestKit:
    def __init__(self, action_cls: type[TimedAction]) -> None:
        self.action_cls = action_cls

    @classmethod
    def of(cls, action_cls: type[TimedAction]) -> TimedActionTestKit:
        return cls(action_cls)

    def call(self, name: str, input: Any = None) -> TimedActionEffect:
        spec = self.action_cls.handlers().get(name)
        if spec is None:
            raise AssertionError(f"{self.action_cls.__name__} has no handler {name!r}")
        input_bytes = spec.input_codec.encode(input) if spec.input_type is not None else b""
        action = self.action_cls(_NoClient())
        return typing.cast(TimedActionEffect, _run(action._run(spec, input_bytes, Metadata())))

File: sdks/typescript/src/testkit/unit.ts
// The unit testkits: one component, no sidecar, no database, no network. Effects come back as values.
// Every input, event, state and reply is encoded and decoded through the component's own codecs on the
// way in and out, so a shape the codec cannot express fails here rather than on first deployment. The
// reduction is the server's (`materialise.ts`), so a test cannot pass on something the sidecar would
// not do.

import { create } from "@bufbuild/protobuf"
import { HttpRequestSchema } from "../_proto/ankka/protocol/v1/endpoint_pb.ts"
import { ComponentClient, noClient } from "../client.ts"
import { binaryCodecs, codecFor, type Codec, type Shape } from "../codec.ts"
import { commandContext, type Principal } from "../context.ts"
import type { ErrorDetail, Metadata, Retention } from "../effects/common.ts"
import type { EventSourcedEffect } from "../effects/eventSourced.ts"
import type { Endpoint, EndpointClass } from "../endpoint.ts"
import { EventSourcedEntity, type EventSourcedEntityClass } from "../eventSourcedEntity.ts"
import type { HandlerRef } from "../handlers.ts"
import { materialiseEventSourced, type Materialised } from "../materialise.ts"
import { Ankka, type RegisteredEventSourced } from "../service.ts"
import { createHttpDispatcher, type HttpDispatcher } from "../server/http.ts"
import type { RouteRef } from "../routes.ts"

export type { Materialised }

function roundTrip<T>(codec: Codec<T>, value: T): T {
  return codec.decode(codec.encode(value))
}

function findHandler(handlers: ReadonlyMap<string, HandlerRef<any, any, any, any>>, handler: HandlerRef<any, any, any, any> | string, who: string): HandlerRef<any, any, any, any> {
  const name = typeof handler === "string" ? handler : handler.name
  const found = handlers.get(name)
  if (!found) throw new Error(`${who} has no handler ${JSON.stringify(name)}; declared: ${[...handlers.keys()].sort().join(", ")}`)
  return found
}

/** Drives one event sourced entity instance through its handlers, folding events as it goes. */
export class EventSourcedTestKit<S, E, C extends EventSourcedEntity<S, E>> {
  readonly #registered: RegisteredEventSourced
  readonly #entity: C
  readonly #client: ComponentClient
  readonly entityId: string
  /** The state now, after every call so far. */
  state: S
  /** The sequence number of the last event, `0n` before any. */
  sequence = 0n
  /** Every event persisted so far, in order. */
  readonly allEvents: E[] = []

  private constructor(cls: EventSourcedEntityClass<S, E, C>, entityId: string, client: ComponentClient) {
    const registry = Ankka.service({ client, log: () => {} })
      .register(cls)
      .validate()
    this.#registered = registry.component(cls.componentId) as RegisteredEventSourced
    this.#entity = new cls()
    this.#entity._bindInstance(entityId)
    this.#client = client
    this.entityId = entityId
    this.state = this.#entity.emptyState()
  }

  /** A kit for `cls` as instance `entityId`. Calls to other components reach `client`, which by default refuses them. */
  static of<S, E, C extends EventSourcedEntity<S, E>>(cls: EventSourcedEntityClass<S, E, C>, entityId = "test", client: ComponentClient = noClient()): EventSourcedTestKit<S, E, C> {
    return new EventSourcedTestKit(cls, entityId, client)
  }

  /** Sends a command or query and returns what its effect amounts to. */
  async call<I, R>(handler: HandlerRef<C, I, R, any> | string, input?: I, metadata: Metadata = {}): Promise<Materialised<S, E, R>> {
    const ref = findHandler(this.#registered.handlers, handler, this.#registered.cls.name)
    const inputCodec = ref.input ? codecFor(ref.input as Shape<I>) : undefined
    const wireInput = inputCodec ? roundTrip(inputCodec, input as I) : undefined
    const context = commandContext(this.#registered.id, this.entityId, this.sequence, metadata)
    this.#entity._bindCommand(this.state, context, this.#client.withMetadata(metadata))
    let effect: EventSourcedEffect<S, E, R>
    try {
      effect = (await ref.run(this.#entity, wireInput)) as EventSourcedEffect<S, E, R>
    } finally {
      this.#entity._unbindCommand()
    }
    if (ref.readOnly && effect.kind !== "read-only") throw new Error(`${this.#registered.id}/${ref.name} is a query and returned a persisting effect`)

    const m = materialiseEventSourced(effect, this.state, (s, e) => this.#entity.applyEvent(s, e))
    const eventCodec = this.#registered.eventCodec as Codec<E>
    const stateCodec = this.#registered.stateCodec as Codec<S>
    const events = m.events.map((e) => roundTrip(eventCodec, e))
    if (m.error) return Object.freeze({ events: [], newState: this.state, retention: m.retention, reply: undefined, error: m.error, noReply: false })
    const newState = roundTrip(stateCodec, m.newState)
    const replyCodec = (ref.reply ? codecFor(ref.reply as Shape<R>) : binaryCodecs.done) as Codec<R>
    const reply = m.noReply ? undefined : roundTrip(replyCodec, m.reply as R)
    this.state = newState
    this.sequence += BigInt(events.length)
    this.allEvents.push(...events)
    return Object.freeze({ events, newState, retention: m.retention, reply, error: undefined, noReply: m.noReply })
  }
}

/** A reply from the endpoint testkit. */
export class Response {
  readonly status: number
  readonly contentType: string
  readonly body: Uint8Array

  constructor(status: number, contentType: string, body: Uint8Array) {
    this.status = status
    this.contentType = contentType
    this.body = body
    Object.freeze(this)
  }

  text(): string {
    return new TextDecoder().decode(this.body)
  }

  json(): unknown {
    return JSON.parse(this.text())
  }
}

export interface RequestOptions {
  readonly query?: Readonly<Record<string, string | readonly string[]>>
  readonly headers?: Readonly<Record<string, string>>
  readonly principal?: Principal
  readonly metadata?: Metadata
}

const PLACEHOLDER = /\{[A-Za-z_][A-Za-z0-9_]*\}/g

/** Calls an endpoint's routes by path, matching them as the sidecar's router does: literal segments outrank parameters. */
export class EndpointTestKit<C extends Endpoint> {
  readonly #cls: EndpointClass<C>
  readonly #dispatcher: HttpDispatcher

  private constructor(cls: EndpointClass<C>, client: ComponentClient) {
    this.#cls = cls
    const registry = Ankka.service({ client, log: () => {} })
      .register(cls)
      .validate()
    this.#dispatcher = createHttpDispatcher({ registry, client, log: () => {} })
  }

  static of<C extends Endpoint>(cls: EndpointClass<C>, client: ComponentClient = noClient()): EndpointTestKit<C> {
    return new EndpointTestKit(cls, client)
  }

  #match(method: string, path: string): { id: string; route: RouteRef<any, any, any, any>; args: string[] } | undefined {
    const prefix = this.#cls.prefix
    const withoutQuery = path.split("?")[0]!
    const rest = withoutQuery.startsWith(prefix) ? withoutQuery.slice(prefix.length) || "/" : withoutQuery
    const candidates: { literals: number; id: string; route: RouteRef<any, any, any, any>; args: string[] }[] = []
    for (const [id, route] of Object.entries(this.#cls.routes)) {
      if (route.method !== method.toUpperCase()) continue
      const pattern = new RegExp("^" + route.template.replace(/[.*+?^$()|[\]\\]/g, "\\$&").replace(PLACEHOLDER, "([^/]+)") + "$")
      const m = pattern.exec(rest)
      if (!m) continue
      const literals = route.template.split("/").filter((seg) => seg && !seg.startsWith("{")).length
      candidates.push({ literals, id, route, args: m.slice(1).map(decodeURIComponent) })
    }
    candidates.sort((a, b) => b.literals - a.literals)
    return candidates[0]
  }

  #request(method: string, path: string, body: unknown, options: RequestOptions) {
    const matched = this.#match(method, path)
    if (!matched) return undefined
    let bytes: Uint8Array = new Uint8Array()
    if (body !== undefined && body !== null) {
      if (body instanceof Uint8Array) bytes = new Uint8Array(body)
      else if (typeof body === "string") bytes = new TextEncoder().encode(body)
      else if (matched.route.body) bytes = (codecFor(matched.route.body) as Codec<unknown>).encode(body)
      else bytes = new TextEncoder().encode(JSON.stringify(body))
    }
    const query: { name: string; value: string }[] = []
    const fromPath = path.includes("?") ? new URLSearchParams(path.slice(path.indexOf("?") + 1)) : undefined
    for (const [name, value] of fromPath ?? []) query.push({ name, value })
    for (const [name, value] of Object.entries(options.query ?? {})) {
      for (const v of typeof value === "string" ? [value] : value) query.push({ name, value: v })
    }
    const headers = Object.entries(options.headers ?? {}).map(([name, value]) => ({ name, value }))
    const principal = options.principal
      ? { subject: options.principal.subject, name: options.principal.name ?? undefined, email: options.principal.email ?? undefined, emailVerified: options.principal.emailVerified, roles: [...options.principal.roles] }
      : undefined
    const req = create(HttpRequestSchema, {
      endpointId: this.#cls.name,
      routeId: matched.id,
      pathArgs: matched.args,
      query,
      headers,
      contentType: matched.route.body ? codecFor(matched.route.body).contentType : "",
      body: bytes,
      principal,
      metadata: { entries: Object.entries(options.metadata ?? {}).map(([key, value]) => ({ key, value })) },
    })
    return { req, route: matched.route }
  }

  async request(method: string, path: string, body?: unknown, options: RequestOptions = {}): Promise<Response> {
    const prepared = this.#request(method, path, body, options)
    if (!prepared) return new Response(404, "text/plain", new TextEncoder().encode(`no route ${method.toUpperCase()} ${path}`))
    if (prepared.route.streaming) {
      const frames: string[] = []
      for await (const f of this.#dispatcher.handleStream(prepared.req)) {
        if (f.frame.case === "text") frames.push(f.frame.value)
        if (f.frame.case === "failed") return new Response(500, "text/plain", new TextEncoder().encode(f.frame.value.message))
      }
      return new Response(200, "text/event-stream", new TextEncoder().encode(frames.join("\n")))
    }
    const reply = await this.#dispatcher.handle(prepared.req)
    if (reply.message.case === "response") {
      const r = reply.message.value
      return new Response(r.status, r.contentType, r.body)
    }
    return new Response(500, "text/plain", new TextEncoder().encode(reply.message.case === "failure" ? reply.message.value.error?.message ?? "failure" : "no reply"))
  }

  /** The frames of an SSE route, each string one frame. */
  async sse(path: string, options: RequestOptions = {}): Promise<string[]> {
    const prepared = this.#request("GET", path, undefined, options)
    if (!prepared) throw new Error(`no route GET ${path}`)
    const frames: string[] = []
    for await (const f of this.#dispatcher.handleStream(prepared.req)) {
      if (f.frame.case === "text") frames.push(f.frame.value)
      if (f.frame.case === "failed") throw new Error(f.frame.value.message)
    }
    return frames
  }

  get(path: string, options?: RequestOptions): Promise<Response> {
    return this.request("GET", path, undefined, options)
  }
  post(path: string, body?: unknown, options?: RequestOptions): Promise<Response> {
    return this.request("POST", path, body, options)
  }
  put(path: string, body?: unknown, options?: RequestOptions): Promise<Response> {
    return this.request("PUT", path, body, options)
  }
  patch(path: string, body?: unknown, options?: RequestOptions): Promise<Response> {
    return this.request("PATCH", path, body, options)
  }
  delete(path: string, body?: unknown, options?: RequestOptions): Promise<Response> {
    return this.request("DELETE", path, body, options)
  }
}

export type { ErrorDetail, Retention }
